source/common/upstream/load_balancer_impl.cc

```diff
@@ -828,8 +857,22 @@ HostConstSharedPtr LeastRequestLoadBalancer::unweightedHostPick(const HostVector
   return candidate_host;
 }
 
+HostConstSharedPtr RandomLoadBalancer::peekAnotherHost(LoadBalancerContext* context) {
+  const absl::optional<HostsSource> hosts_source = hostSourceToUse(context, random(true));
+  if (!hosts_source) {
+    return nullptr;
+  }
+
+  const HostVector& hosts_to_use = hostSourceToHosts(*hosts_source);
+  if (hosts_to_use.empty()) {
+    return nullptr;
+  }
+
+  return hosts_to_use[random(true) % hosts_to_use.size()];
```

Comment:
the 2 random calls in peekAnotherHost are interesting.  It does require that this method and RandomLoadBalancer::chooseHostOnce remain in sync.  It may make sense to use a helper function to implement both methods which takes isPrefetching as an argument.

The other oddity are the early returns in this method; if that happens only 1 random number is added per call instead of 2.  I wonder if we should stashed_random_.pop_back() when handling an early return.

Another question: should we consider using the same random number as the hash for hostSourceToUse() and the vector lookup in this line?